File: src/protocol.rs
use fspiox_api::*;
use mojaloop_api::central_ledger::{participants, settlement_models};
use mojaloop_api::settlement::{settlement, settlement_windows};
use serde::{Deserialize, Serialize};

#[cfg(feature = "typescript_types")]
use ts_rs::{TS, export};

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct TransferMessage {
    pub msg_sender: common::FspId,
    pub msg_recipient: common::FspId,
    pub currency: common::Currency,
    pub amount: common::Amount,
    // If we have the caller provide the transfer ID, we can use this as a unique message reference
    // for this message sequence with this caller.
    pub transfer_id: transfer::TransferId,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
pub struct AccountInitialization {
    pub currency: common::Currency,
    pub initial_position: common::Amount,
    pub ndc: u32,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SettlementWindowCloseMessage {
    pub id: settlement_windows::SettlementWindowId,
    pub reason: String,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum ClientMessage {
    /// Run end-to-end transfers
    Transfers(Vec<TransferMessage>),
    /// Create hub settlement and reconciliation accounts
    CreateHubAccounts(Vec<participants::HubAccount>),
    // TODO: this _could_ be a vector of vectors of accounts. Each 0th-level vector would represent
    // a participant, and each 1st-level vector would contain desired accounts.
    // TODO: disable the participants on socket closure
    /// Create a set of participants. Will be disabled when the socket disconnects.
    CreateParticipants(Vec<AccountInitialization>),
    /// Create a settlement model
    CreateSettlementModel(settlement_models::SettlementModel),
    /// Attempt to close the currently open settlement window. Will fail if the window does not
    /// contain any transfers.
    CloseSettlementWindow(SettlementWindowCloseMessage),
    /// Get settlement windows
    GetSettlementWindows(settlement_windows::GetSettlementWindows),
    /// Get settlements
    GetSettlements(settlement::GetSettlements),
    // /// Generate some closed settlement windows with the given transfers. Will close the currently
    // /// open settlement window if that contains any existing transfers.
    // CreateSettlementWindows(Vec<Vec<TransferMessage>>),
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct SettlementModelCreatedMessage {
    pub settlement_model: settlement_models::SettlementModel,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct TransferCompleteMessage {
    pub id: transfer::TransferId,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct TransferErrorMessage {
    pub id: transfer::TransferId,
    pub response: common::ErrorResponse,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct ClientParticipant {
    pub name: common::FspId,
    pub account: AccountInitialization,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
pub struct SettlementWindowCloseFailedMessage {
    pub id: settlement_windows::SettlementWindowId,
    pub response: fspiox_api::common::ErrorResponse,
}

#[cfg_attr(feature = "typescript_types", derive(TS))]
#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type", content = "value")]
pub enum ServerMessage {
    TransferComplete(TransferCompleteMessage),
    TransferError(TransferErrorMessage),
    AssignParticipants(Vec<ClientParticipant>),
    HubAccountsCreated(Vec<participants::HubAccount>),
    SettlementModelCreated(SettlementModelCreatedMessage),
    SettlementWindowClosed(settlement_windows::SettlementWindowId),
    SettlementWindowCloseFailed(SettlementWindowCloseFailedMessage),
    SettlementWindows(Vec<settlement_windows::SettlementWindow>),
    Settlements(settlement::Settlements),
}

#[cfg(feature = "typescript_types")]
export! {
    TransferCompleteMessage,
    TransferErrorMessage,
    AccountInitialization,
    ClientParticipant,
    TransferMessage,
    SettlementModelCreatedMessage,
    SettlementWindowCloseMessage,
    SettlementWindowCloseFailedMessage,
    settlement_windows::SettlementWindowState,
    settlement_windows::SettlementWindowContent,
    settlement_windows::SettlementWindow,
    settlement_windows::SettlementWindowId,
    settlement_windows::SettlementWindowContentId,
    settlement_windows::GetSettlementWindows,
    settlement_models::SettlementAccountType,
    settlement_models::SettlementDelay,
    settlement_models::SettlementGranularity,
    settlement_models::SettlementInterchange,
    settlement_models::LedgerAccountType,
    settlement_models::SettlementModel,
    settlement::GetSettlements,
    settlement::Settlement,
    settlement::SettlementId,
    settlement::SettlementState,
    settlement::SettlementParticipant,
    settlement::SettlementAccount,
    settlement::ParticipantId,
    settlement::ParticipantCurrencyId,
    settlement::NetSettlementAmount,
    participants::HubAccount,
    participants::HubAccountType,
    common::FspId,
    common::DateTime,
    common::Amount,
    common::Currency,
    common::ErrorResponse,
    common::ErrorInformation,
    common::MojaloopApiError,
    ServerMessage,
    ClientMessage => "clients/typescript/src/lib/protocol.ts"
}

File: clients/typescript/src/lib/protocol.ts
export interface TransferCompleteMessage {
  id: string;
}

export interface TransferErrorMessage {
  id: string;
  response: ErrorResponse;
}

export interface AccountInitialization {
  currency: Currency;
  initial_position: Amount;
  ndc: number;
}

export interface ClientParticipant {
  name: FspId;
  account: AccountInitialization;
}

export interface TransferMessage {
  msg_sender: FspId;
  msg_recipient: FspId;
  currency: Currency;
  amount: Amount;
  transfer_id: string;
}

export interface SettlementModelCreatedMessage {
  settlement_model: SettlementModel;
}

export interface SettlementWindowCloseMessage {
  id: SettlementWindowId;
  reason: string;
}

export interface SettlementWindowCloseFailedMessage {
  id: SettlementWindowId;
  response: ErrorResponse;
}

export type SettlementWindowState =
  | "OPEN"
  | "CLOSED"
  | "PENDING_SETTLEMENT"
  | "SETTLED"
  | "ABORTED";

export interface SettlementWindowContent {
  id: SettlementWindowContentId;
  settlementWindowId: SettlementWindowId | null;
  state: SettlementWindowState;
  ledgerAccountType: LedgerAccountType;
  currencyId: Currency;
  createdDate: DateTime;
  changedDate: DateTime | null;
  settlementId: SettlementId | null;
}

export interface SettlementWindow {
  settlementWindowId: SettlementWindowId;
  reason: string | null;
  state: SettlementWindowState;
  createdDate: DateTime;
  changedDate: DateTime | null;
  content: SettlementWindowContent[] | null;
}

export type SettlementWindowId = number;

export type SettlementWindowContentId = number;

export interface GetSettlementWindows {
  currency: Currency | null;
  participantId: FspId | null;
  state: SettlementWindowState | null;
  fromDateTime: DateTime | null;
  toDateTime: DateTime | null;
}

export type SettlementAccountType = "SETTLEMENT" | "INTERCHANGE_FEE_SETTLEMENT";

export type SettlementDelay = "DEFERRED" | "IMMEDIATE";

export type SettlementGranularity = "GROSS" | "NET";

export type SettlementInterchange = "BILATERAL" | "MULTILATERAL";

export type LedgerAccountType = "INTERCHANGE_FEE" | "POSITION";

export interface SettlementModel {
  autoPositionReset: boolean;
  ledgerAccountType: LedgerAccountType;
  settlementAccountType: SettlementAccountType;
  name: string;
  requireLiquidityCheck: boolean;
  settlementDelay: SettlementDelay;
  settlementGranularity: SettlementGranularity;
  settlementInterchange: SettlementInterchange;
  currency: Currency;
}

export interface GetSettlements {
  currency: Currency | null;
  participantId: FspId | null;
  settlementWindowId: SettlementWindowId | null;
  state: SettlementState | null;
  fromDateTime: DateTime | null;
  toDateTime: DateTime | null;
  fromSettlementWindowDateTime: DateTime | null;
  toSettlementWindowDateTime: DateTime | null;
}

export interface NewSettlement {
  settlementModel: string;
  reason: string;
  settlementWindows: WindowParametersNewSettlement[];
}

export interface SettlementSettlementWindow {
  id: SettlementWindowId;
  reason: string | null;
  state: SettlementWindowState;
  createdDate: DateTime;
  changedDate: DateTime | null;
  content: SettlementWindowContent[] | null;
}

export interface Settlement {
  id: SettlementId;
  state: SettlementState;
  createdDate: DateTime;
  changedDate: DateTime;
  settlementWindows: SettlementSettlementWindow[];
  participants: SettlementParticipant[];
}

export type SettlementId = number;

export type SettlementState =
  | "PENDING_SETTLEMENT"
  | "PS_TRANSFERS_RECORDED"
  | "PS_TRANSFERS_RESERVED"
  | "PS_TRANSFERS_COMMITTED"
  | "SETTLING"
  | "SETTLED"
  | "ABORTED";

export interface SettlementParticipant {
  id: ParticipantId;
  accounts: SettlementAccount[];
}

export interface SettlementAccount {
  id: ParticipantCurrencyId;
  reason: string;
  state: SettlementState;
  netSettlementAmount: NetSettlementAmount;
}

export type ParticipantId = number;

export type ParticipantCurrencyId = number;

export interface NetSettlementAmount {
  amount: Amount;
  currency: Currency;
}

export interface WindowParametersNewSettlement {
  id: SettlementWindowId;
}

export interface HubAccount {
  type: HubAccountType;
  currency: Currency;
}

export type HubAccountType =
  | "HUB_MULTILATERAL_SETTLEMENT"
  | "HUB_RECONCILIATION";

export type FspId = string;

export type DateTime = string;

export type Amount = string;

export type Currency =
  | "AED"
  | "AFA"
  | "AFN"
  | "ALL"
  | "AMD"
  | "ANG"
  | "AOA"
  | "AOR"
  | "ARS"
  | "AUD"
  | "AWG"
  | "AZN"
  | "BAM"
  | "BBD"
  | "BDT"
  | "BGN"
  | "BHD"
  | "BIF"
  | "BMD"
  | "BND"
  | "BOB"
  | "BOV"
  | "BRL"
  | "BSD"
  | "BTN"
  | "BWP"
  | "BYN"
  | "BYR"
  | "BZD"
  | "CAD"
  | "CDF"
  | "CHE"
  | "CHF"
  | "CHW"
  | "CLF"
  | "CLP"
  | "CNY"
  | "COP"
  | "COU"
  | "CRC"
  | "CUC"
  | "CUP"
  | "CVE"
  | "CZK"
  | "DJF"
  | "DKK"
  | "DOP"
  | "DZD"
  | "EEK"
  | "EGP"
  | "ERN"
  | "ETB"
  | "EUR"
  | "FJD"
  | "FKP"
  | "GBP"
  | "GEL"
  | "GGP"
  | "GHS"
  | "GIP"
  | "GMD"
  | "GNF"
  | "GTQ"
  | "GYD"
  | "HKD"
  | "HNL"
  | "HRK"
  | "HTG"
  | "HUF"
  | "IDR"
  | "ILS"
  | "IMP"
  | "INR"
  | "IQD"
  | "IRR"
  | "ISK"
  | "JEP"
  | "JMD"
  | "JOD"
  | "JPY"
  | "KES"
  | "KGS"
  | "KHR"
  | "KMF"
  | "KPW"
  | "KRW"
  | "KWD"
  | "KYD"
  | "KZT"
  | "LAK"
  | "LBP"
  | "LKR"
  | "LRD"
  | "LSL"
  | "LTL"
  | "LVL"
  | "LYD"
  | "MAD"
  | "MDL"
  | "MGA"
  | "MKD"
  | "MMK"
  | "MNT"
  | "MOP"
  | "MRO"
  | "MUR"
  | "MVR"
  | "MWK"
  | "MXN"
  | "MXV"
  | "MYR"
  | "MZN"
  | "NAD"
  | "NGN"
  | "NIO"
  | "NOK"
  | "NPR"
  | "NZD"
  | "OMR"
  | "PAB"
  | "PEN"
  | "PGK"
  | "PHP"
  | "PKR"
  | "PLN"
  | "PYG"
  | "QAR"
  | "RON"
  | "RSD"
  | "RUB"
  | "RWF"
  | "SAR"
  | "SBD"
  | "SCR"
  | "SDG"
  | "SEK"
  | "SGD"
  | "SHP"
  | "SLL"
  | "SOS"
  | "SPL"
  | "SRD"
  | "SSP"
  | "STD"
  | "SVC"
  | "SYP"
  | "SZL"
  | "THB"
  | "TJS"
  | "TMT"
  | "TND"
  | "TOP"
  | "TRY"
  | "TTD"
  | "TVD"
  | "TWD"
  | "TZS"
  | "UAH"
  | "UGX"
  | "USD"
  | "USN"
  | "UYI"
  | "UYU"
  | "UZS"
  | "VEF"
  | "VND"
  | "VUV"
  | "WST"
  | "XAF"
  | "XAG"
  | "XAU"
  | "XCD"
  | "XDR"
  | "XFO"
  | "XFU"
  | "XOF"
  | "XPD"
  | "XPF"
  | "XPT"
  | "XSU"
  | "XTS"
  | "XUA"
  | "XXX"
  | "YER"
  | "ZAR"
  | "ZMK"
  | "ZMW"
  | "ZWD"
  | "ZWL"
  | "ZWN"
  | "ZWR";

export interface ErrorResponse {
  errorInformation: ErrorInformation;
}

export interface ErrorInformation {
  errorCode: MojaloopApiError;
  errorDescription: string;
}

export type MojaloopApiError =
  | "1000"
  | "1001"
  | "2000"
  | "2001"
  | "2002"
  | "2003"
  | "2004"
  | "2005"
  | "3000"
  | "3000"
  | "3001"
  | "3002"
  | "3003"
  | "3040"
  | "3100"
  | "3101"
  | "3102"
  | "3103"
  | "3104"
  | "3105"
  | "3106"
  | "3107"
  | "3200"
  | "3201"
  | "3202"
  | "3203"
  | "3204"
  | "3205"
  | "3206"
  | "3207"
  | "3208"
  | "3209"
  | "3210"
  | "3300"
  | "3301"
  | "3302"
  | "3303"
  | "4000"
  | "4001"
  | "4100"
  | "4101"
  | "4102"
  | "4103"
  | "4200"
  | "4300"
  | "4400"
  | "5000"
  | "5001"
  | "5100"
  | "5101"
  | "5102"
  | "5103"
  | "5104"
  | "5105"
  | "5106"
  | "5200"
  | "5300"
  | "5400";

export type ServerMessage =
  | { type: "TransferComplete"; value: TransferCompleteMessage }
  | { type: "TransferError"; value: TransferErrorMessage }
  | { type: "AssignParticipants"; value: ClientParticipant[] }
  | { type: "HubAccountsCreated"; value: HubAccount[] }
  | { type: "SettlementModelCreated"; value: SettlementModelCreatedMessage }
  | { type: "SettlementWindowClosed"; value: SettlementWindowId }
  | {
    type: "SettlementWindowCloseFailed";
    value: SettlementWindowCloseFailedMessage;
  }
  | { type: "SettlementWindows"; value: SettlementWindow[] }
  | { type: "Settlements"; value: Settlement[] }
  | { type: "NewSettlementCreated"; value: Settlement };

export type ClientMessage =
  | { type: "Transfers"; value: TransferMessage[] }
  | { type: "CreateHubAccounts"; value: HubAccount[] }
  | { type: "CreateParticipants"; value: AccountInitialization[] }
  | { type: "CreateSettlementModel"; value: SettlementModel }
  | { type: "CloseSettlementWindow"; value: SettlementWindowCloseMessage }
  | { type: "GetSettlementWindows"; value: GetSettlementWindows }
  | { type: "GetSettlements"; value: GetSettlements }
  | { type: "CreateSettlement"; value: NewSettlement };